deals/module/deals.go

```diff
@@ -235,61 +235,125 @@ func (m *Module) Watch(ctx context.Context, proposals []cid.Cid) (<-chan deals.D
+	c := deals.ListDealRecordsConfig{}
+	for _, opt := range opts {
+		opt(&c)
+	}
+	var final []deals.StorageDealRecord
+	if !c.OnlyPending {
+		recs, err := m.store.getFinalDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting final deals: %v", err)
+		}
+		final = recs
 	}
-	return ret, nil
-}
 
-// PendingDealRecords returns a list of all pending storage deals.
-// Records are sorted ascending by timestamp.
-func (m *Module) PendingDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	var pending []deals.StorageDealRecord
+	if c.IncludePending || c.OnlyPending {
+		recs, err := m.store.getPendingDeals()
+		if err != nil {
+			return nil, fmt.Errorf("getting pending deals: %v", err)
+		}
+		pending = recs
 	}
-	return ret, nil
-}
 
-// AllDealRecords returns a list of all finalized and pending deals.
-// Records are sorted ascending by activation epoch, if available, then timestamp.
-func (m *Module) AllDealRecords() ([]deals.DealRecord, error) {
-	ret, err := m.store.getFinalDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting final deals: %v", err)
-	}
-	pending, err := m.store.getPendingDeals()
-	if err != nil {
-		return nil, fmt.Errorf("getting pending deals: %v", err)
+	combined := append(final, pending...)
+
+	var filtered []deals.StorageDealRecord
+
+	if len(c.FromAddrs) > 0 || len(c.DataCids) > 0 {
+		var fromAddrsFilter map[string]struct{}
+		var dataCidsFilter map[string]struct{}
+		for _, addr := range c.FromAddrs {
+			fromAddrsFilter[addr] = struct{}{}
+		}
+		for _, cid := range c.DataCids {
+			dataCidsFilter[cid] = struct{}{}
+		}
+		for _, record := range combined {
+			_, inFromAddrsFilter := fromAddrsFilter[record.Addr]
+			_, inDataCidsFilter := dataCidsFilter[record.DealInfo.PieceCID.String()]
+			includeViaFromAddrs := len(c.FromAddrs) == 0 || inFromAddrsFilter
+			includeViaDataCids := len(dataCidsFilter) == 0 || inDataCidsFilter
+			if includeViaFromAddrs && includeViaDataCids {
+				filtered = append(filtered, record)
+			}
+		}
+	} else {
+		filtered = combined
 	}
-	ret = append(ret, pending...)
-	sort.Slice(ret, func(i, j int) bool {
-		if ret[i].Pending || ret[j].Pending {
-			return ret[i].Time < ret[j].Time
+
+	sort.Slice(filtered, func(i, j int) bool {
+		l := filtered[j]
+		r := filtered[i]
+		if c.Ascending {
+			l = filtered[i]
+			r = filtered[j]
+		}
+		if l.Pending || r.Pending {
+			return l.Time < r.Time
 		}
-		if ret[i].DealInfo.ActivationEpoch < ret[j].DealInfo.ActivationEpoch {
+		if l.DealInfo.ActivationEpoch < r.DealInfo.ActivationEpoch {
 			return true
 		}
-		if ret[i].DealInfo.ActivationEpoch > ret[j].DealInfo.ActivationEpoch {
+		if l.DealInfo.ActivationEpoch > r.DealInfo.ActivationEpoch {
 			return false
 		}
-		return ret[i].Time < ret[j].Time
+		return l.Time < r.Time
```

Comment:
Sort by timestamp if either record is pending, activation epoch otherwise, use timestamp as secondary sort when activation epochs are the same.